### ISSUE TITLE
16.2-Compression: else

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#84]
# Knowledge Distillation
Distillation = 蒸餾

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/d7d00062-ca6b-401f-9ce5-b2ce8253220f)

先 Train 一個大的 Network 在 Knowledge Distillation 裡面 叫做 Teacher Network 
你真正想要的那個小的 Network 叫做 Student Network

你先 Train 一個大的 Network 叫做 Teacher Network 
再根據這個大的 Network來製造 Student Network

假設要做手寫數字辨識
把你的訓練資料 都丟到 Teacher 裡面 
然後 Teacher 產生 Output = 一個 Distribution
看到這張圖片 1 的分數是 0.7 
看到這張圖片 7 的分數是 0.2
看到這張圖片 9 的分數是 0.1

學生不是去看這個圖片的正確答案來學習
它把老師的輸出 就當做正確答案 
也就是老師輸出  
1 要 0.7
7 要 0.2 
9 要 0.1 
那學生的輸出呢 也就要儘量去逼近老師的輸出
儘量去逼近 1 是 0.7 7 是 0.2 9 是 0.1 這樣的答案
學生就是根據老師的答案學 就算老師的答案是錯的 學生就去學一個錯的東西

Teacher Network 其實會提供 Student Network 額外的資訊 
老師說 1 是 0.7 7 是 0.2
學生只要學到 1 是 0.7 7 是 0.2 就夠了 
這樣反而可以讓小的 Network學得比直接根據正確的答案要學來得要好

Teacher Network 不一定要是單一一個巨大的 Network 它甚至可以是多個 Network 的 Ensemble
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/35cbf349-f059-4937-9f01-656d10ca650a)

Ensemble 訓練多個模型 輸出 多個模型投票的結果或者是多個模型的輸出平均起來的結果
在比賽裡面常常會使用到 Ensemble 的方法, 但計算量大很難用在實際系統上

要用在實際的系統上可以把多個 Ensemble 起來的 Network綜合起來變成一個 
Knowledge Distillation:把多個 Network Ensemble 起來的結果 當做是 Teacher Network 的輸出

### Temperature for softmax 
在使用 Knowledge Distillation 的時候會稍微改一下 Softmax 的 Function
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/6072d66c-c03b-483e-b23c-f3ae2470a107)

Softmax Softmax :把每一個 Neural 的輸出 都去 Exponential 然後再做 Normalize
可以讓你的 Network 的輸出 變成一個機率的分布 分子都是介於 0 到 1 之間的

Temperature 是一個 Hyperparameters T 呢
在做 Exponential 之前 把每一個數值呢都除上一個 T

y都趨近於 0 跟直接和正確的答案學幾乎沒有不同
跟 Teacher 學的一個好處就是老師會告訴你說 哪些類別其實是比較像的
讓 Student Network 在學的時候不會那麼辛苦

_________________

### Parameter Quantization(作業是用不上)

用比較少的空間來儲存一個參數
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/24577004-0312-40f4-9f9f-75dd0aa286a9)

你可能是用 64 個 Bits  但會不會用 16 個 Bits 就夠了
你的儲存空間  (Network 的大小) 直接就砍掉很多
而且 Performance 不會掉很多 甚至有時候還會稍微更好一點

Weight Clustering Weight Clustering 更進一步壓縮參數的方法
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/6632c373-6f69-4d55-be29-93e62f325a22)

Network 的參數做分群
接下來每一群只拿一個數值來表示它
這樣做的好處是 儲存你的參數的時候 你就只要記兩個東西 
一個是一個表格 另外一個是屬於哪一群

你其實還可以把參數再更進一步做壓縮-Huffman Encoding
最終極的結果就是 你可以只拿一個 Bits 來存每一個參數
你可以說你的 Network 裡面的 Weight 要嘛就是 +1 要嘛就是 -1
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/236f7acd-3893-4db8-8346-8c59a9e5d8ec)
Binary Network 的 厲害的地方是他甚至正確率還有可能更好
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/e455c4ba-154f-4fc0-93f6-a50a99856e3f)

_____

Architecture Design

透過 Network 架構的設計來達到減少參數量的效果

複習 CNN
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/18a39254-a13e-428d-b5ff-e26f1fc474e2)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/e6930a5a-e0d5-45df-b6dc-4c8234eef198)
有幾個 Filter 那 Output Feature Map 就有幾個 Channel
這邊有 4 個 Filter 那 Output Feature Map 就有 4 個 Channel

### Depthwise Separable Convolution
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/93733ab4-8627-4455-ba97-02f86bc91d2a)

Depthwise Separable 的 Convolution 分成兩個步驟
1. Depthwise Convolution 
2. Pointwise Convolution 

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/3a05c92e-cfbb-445f-99c6-b292b84e8c01)

1. Depthwise Convolution 
有n個 Channel 就有n個 Filter 而每一個 Filter 只管一個 Channel
假設 Input Feature Map 是兩個 Channel 
那在 Depthwise 的 Convolution Layer 裡面我們就只放兩個 Filter 

在 Depthwise Convolution 裡面 你 Input 有幾個 Channel
你 Output 的 Channel 的數目會是一模一樣的

一般的 Convolution Layer 裡面 Input 跟 Output 的 Channel 數目可以不一樣
但在 Depthwise Convolution 裡面Input 跟 Output 的 Channel 數目是一模一樣的

2. Pointwise Convolution 
Channel 和 Channel 之間的互動

我們現在一樣有一堆 Filter
這個跟一般的 Convolution Layer 是一樣的
但限制是 Filter 的 Kernel 的 Size 通通都是 1 x 1 

Pointwise Pointwise Convolution跟一般的 Convolution Layer 一樣的地方是
輸入跟輸出的 Channel 數目可以不一樣
但是 Pointwise Convolution Filter 的大小只準是 1 x 1
你只要考慮 Channel 之間的關係就好了
你就不要考慮同一個 Channel 內部的關係了
Depthwise Convolution 已經做完了同一個 Channel 內部的關係

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/947c7202-2fa2-4aef-88b4-8b2e088af596)

因為 O 通常是一個很大的值( Channel 數目可能開個 256 or 512)
所以先把這個 1 除以 O 忽略不計

Depthwise Separable Convolution 來自於Low Rank Approximation

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/d44c0f9e-06b2-4a7c-a5a7-60a4db0e554f)

假設你有某一個 Layer 
輸入有 N 個 Neuron
輸出有 M 個 Neuron 
N 或者是 M 其中一者非常大
這兩層之間的參數量是 N x M 
只要 N 跟 M 其中一者很大  W 的參數就很多了

在 N 跟 M 中間再插一層
第一層呢用 V 來表示
第二層用 U 來表示

這個兩層的 Network 參數量變少
第一層是 N x K
第二層是 K x M

當 K 遠小於 M 跟 N
U 跟 V 的參數量加起來是比 W 還要少的多
整體而言只要可以 K 夠小 U + V 的參數量就會變少

雖然可以減少參數量  但你會減少 W 的可能性

其實Depthwise Separable Convolution用的就是Low Rank Approximation的概念
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/39363dbe-6f40-4eda-a05c-c6f9831a6487)

第一個 Channel 左上角這 9 個值 跟第二個 Channel 左上角這 9 個值
可以看成是 我們有兩個 Filter這兩個 Filter 它分別是 9 個 Input 
然後接下來這兩個 Filter 的輸出再把它綜合起來得到最終的輸出

本來是直接從這 18 個數值變成一個數值
現在是分兩階 18 個數值變兩個數值再變一個數值 

________

### Dynamic Computation

希望 Network 可以自由地調整它需要的運算量
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/8eb4d97e-b327-4437-8104-63227202a020)

讓 Network 自由地調整它對運算資源的需求方法之一 :機器自由地調整Network 的深度
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/15d78b7a-cd0f-4fd2-be84-0cc3c173cab2)

讓 Network 自由地調整它對運算資源的需求方法之一 :機器自由地調整Network 的寬度
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/8ccaece8-48dc-49f8-9cc8-b0db69ca72cf)

我們可以自由去決定它的深度跟它的寬度
但是所謂的決定權還是在人這一邊

但Network 是有辦法自己去決定它的寬度跟深度的
